activerecord/lib/active_record/connection_adapters.rb

```diff
@@ -31,6 +31,68 @@ def resolve(adapter_name) # :nodoc:
         class_name, path_to_adapter = @adapters[adapter_name.to_s]
 
         unless class_name
+          # To provide better error messages for adapters expecting the pre-7.2 adapter registration API, we attempt
+          # to load the adapter file from the old location which was required by convention, and then raise an error
+          # describing how to upgrade the adapter to the new API.
+          legacy_adapter_path = "active_record/connection_adapters/#{adapter_name}_adapter"
+          legacy_adapter_connection_method_name = "#{adapter_name}_connection".to_sym
+          legacy_adapter_class_method_name = "#{adapter_name}_adapter_class".to_sym # sqlite3_adapter_class
+
+          begin
+            # Attempt to load the legacy adapter by the filename convention.
+            require legacy_adapter_path
+
+            # If it is loaded without raising we found the found a file that may be the legacy adapter.
+            if ActiveRecord::ConnectionHandling.method_defined?(legacy_adapter_connection_method_name) &&
+              ActiveRecord::ConnectionHandling.method_defined?(legacy_adapter_class_method_name) &&
+              ActiveRecord::Base.respond_to?(legacy_adapter_class_method_name)
+              # If we find the connection method then we care certain it is a legacy adapter. We can force the legacy
+              # adapter to load by calling the legacy methods, flag it as deprecated, and remove this path and exception
+              # in the subsequent version.
+              legacy_adapter_class = ActiveRecord::Base.public_send(legacy_adapter_class_method_name)
+
+              # Register the adapter with the new API so that we don't break backwards compatibility in this version.
+              register(adapter_name, legacy_adapter_class.name, legacy_adapter_path)
```

Comment:
This assumes that `<adapter>_connection` (the legacy method) simply call `Adapter.new` but that isn't always true.

If you look at my original PR that introduced register, you'll see that one adapter still had some extra logic in `_connection` that I had to move to `initialize`.

So we can't assume 3rd party adapter don't have a similar problem. 

It's probably best to keep using the old method, and not try to register the legacy adapter. We might want however to make sure we don't spam that deprecation every time a connection is created.

Reply:
Nothing is every easy is it.